code/datums/components/proximity_monitor.dm

```diff
@@ -0,0 +1,140 @@
+	proximity_checkers = list()
+	for(var/turf/T in range(1, get_turf(parent)))
+		var/obj/effect/abstract/proximity_checker/P = new(T, parent)
+		proximity_checkers += P
+		// Basic movement for the proximity_checker objects. The objects will move 1 tile in the direction the parent just moved.
+		P.RegisterSignal(parent, COMSIG_MOVABLE_MOVED, /obj/effect/abstract/proximity_checker/.proc/HandleMove)
+
+/**
+ * Re-centers all of the parent's `proximity_checker`s around its current location.
+ */
+/datum/component/proximity_monitor/proc/recenter_prox_checkers()
+	var/list/prox_checkers = owner.get_all_adjacent_turfs()
+	for(var/checker in proximity_checkers)
+		var/obj/effect/abstract/proximity_checker/P = checker
+		P.loc = pick_n_take(prox_checkers)
+
+/**
+ * # Proximity checker abstract object
+ *
+ * Inteded for use with the proximity checker component (/datum/component/proximity_monitor).
+ * Whenever a movable atom crosses this object, it calls `HasProximity()` on the object which is listening for proximity (`hasprox_receiver`).
+ */
+/obj/effect/abstract/proximity_checker
+	name = "Proximity checker"
+	/// Whether or not the proximity checker is listening for things crossing it.
+	var/active
+	/// The linked atom which has the proximity_monitor component, and will recieve the `HasProximity()` calls.
+	var/atom/hasprox_receiver
+
+// If this object is initialized without a `_hasprox_receiver` arg, it is qdel'd.
+/obj/effect/abstract/proximity_checker/Initialize(mapload, atom/_hasprox_receiver)
+	if(_hasprox_receiver)
+		hasprox_receiver = _hasprox_receiver
+		RegisterSignal(hasprox_receiver, COMSIG_PARENT_QDELETING, .proc/OnParentDeletion)
+		if(isturf(hasprox_receiver.loc)) // if the reciever is inside a locker/crate/etc, they don't detect proximity
+			active = TRUE
+	else
+		stack_trace("/obj/effect/abstract/proximity_checker created without a receiver")
+		return INITIALIZE_HINT_QDEL
+	return ..()
+
+/obj/effect/abstract/proximity_checker/Destroy()
+	hasprox_receiver = null
+	return ..()
+
+/**
+ * Called when the `hasprox_receiver` receives the `COMSIG_PARENT_QDELETING` signal. When the receiver is deleted, so is this object.
+ *
+ * Arugments:
+ * * source - this will be the `hasprox_receiver`
+ * * force - the force flag taken from the qdel proc currently running on `hasprox_receiver`
+ */
+/obj/effect/abstract/proximity_checker/proc/OnParentDeletion(datum/source, force = FALSE)
+	qdel(src)
+
+/**
+ * Something crossed over the proximity_checker. Notify the `hasprox_receiver` it has proximity with something. Only fires if the checker is `active`.
+ */
+/obj/effect/abstract/proximity_checker/Crossed(atom/movable/AM, oldloc)
+	set waitfor = FALSE
+	if(active)
+		hasprox_receiver.HasProximity(AM)
+
+/**
+ * Moves the proximity_checker 1 tile in the `Dir` direction. If `Dir` is null, it doesn't change location.
+ *
+ * If the new location of the receiver is NOT a turf, set `active` to FALSE, so that it does not receive proximity calls.
```

Comment:
It suggests that this is part of the logic whether or not Dir is null. Which is not the case since if Dir is not null it won't do this

Reply:
Ah you are referring to the fact that it will teleport if dir is null right?

Reply:
Nah meant that it will return if Dir is not null. Thus not set the `active` variable in the same proc call

Reply:
Ah gotcha

Reply:
I just realized I have it return for a reason. If `Dir` is not null, its just a standard movement in 1 tile. It *will* be on a turf, therefore I don't want to keep checking if its on a turf or not.